bench/rttanalysis: shard TestBenchmarkExpectation to avoid timeouts

spilchen opened the pull request:
The TestBenchmarkExpectation benchmark has been frequently timing out after 15 minutes.  This appears to be caused by slow CI machines rather than issues with the test logic itself.

To address this, the test is now split into four shards. Each shard is executed separately and receives the full 15-minute timeout budget. This should reduce the likelihood of timeout test failures.

Fixes #148384

Release note: none
Epic: none

rafiss reviewed (approved):
i like this idea! just had minor comments

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153721#-:-O_TXBwe7xmDnGcQFJsB:b1iel7h)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/bench/rttanalysis/validate_benchmark_data_test.go` line 14 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153721#-O_TX4MZDh577cqjlqHa:-O_TX4MZDh577cqjlqHb:b-3iqu4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/eed425db4cbd8a0524a07d9ca43bc0ccbfd089f1/pkg/bench/rttanalysis/validate_benchmark_data_test.go#L14)):*
> ```Go
> 
> func TestBenchmarkExpectationShard1(t *testing.T) {
> 	reg.RunExpectations(t, 1, 4)
> ```

super nit: let's make a `const` for shardCount=4
___
*[`pkg/bench/rttanalysis/registry.go` line 84 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153721#-O_TWc6R7Vh0_yfEQo8U:-O_TWc6R7Vh0_yfEQo8V:b-c2n26l) ([raw file](https://github.com/cockroachdb/cockroach/blob/eed425db4cbd8a0524a07d9ca43bc0ccbfd089f1/pkg/bench/rttanalysis/registry.go#L84)):*
> ```Go
> 		// Distribute test groups across shards using hash-based assignment
> 		for groupName, testCases := range r.r {
> 			h := fnv32Hash(groupName)
> ```

i'm curious if there are enough names in the registry to actually get an even number of tests across the shards? would we be able to get away with using round-robin here? (i wasn't sure why we needed to have consistent shard assignment)


<!-- Sent from Reviewable.io -->

rickystewart reviewed (approved):
All of my comments are nits and could be ignored.

spilchen commented on pkg/bench/rttanalysis/validate_benchmark_data_test.go:
```diff
@@ -5,14 +5,23 @@
-	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
-)
+// NOTE: If you change the number of shards, you must also update the
+// shard_count in BUILD.bazel to match.
```
Good idea
<!-- Reviewable comment -O_WxUYDC7KdTkmde9dx:b-ul8jrf -->
<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153721)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/bench/rttanalysis/registry.go` line 84 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153721#-O_TWc6R7Vh0_yfEQo8U:-O_Wv_93E-jXF7toPLEQ:b-qdipmm) ([raw file](https://github.com/cockroachdb/cockroach/blob/eed425db4cbd8a0524a07d9ca43bc0ccbfd089f1/pkg/bench/rttanalysis/registry.go#L84)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i'm curious if there are enough names in the registry to actually get an even number of tests across the shards? would we be able to get away with using round-robin here? (i wasn't sure why we needed to have consistent shard assignment)
</blockquote></details>

There are 31 names in the registry. With the hash assignment, each shard ran: 6, 10, 4, 10. Not too bad, but the 3rd shard was under represented. It seems like the names in the registry are fairly stable, so I don't know if there is a need for consistent hashing. I changed to the round robin approach you mentioned. 
___
*[`pkg/bench/rttanalysis/validate_benchmark_data_test.go` line 14 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153721#-O_TX4MZDh577cqjlqHa:-O_Wv0Wp-kWNVegKYcfy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eed425db4cbd8a0524a07d9ca43bc0ccbfd089f1/pkg/bench/rttanalysis/validate_benchmark_data_test.go#L14)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: let's make a `const` for shardCount=4
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

spilchen committed:
bench/rttanalysis: shard TestBenchmarkExpectation to avoid timeouts

The TestBenchmarkExpectation benchmark has been frequently timing out
after 15 minutes.  This appears to be caused by slow CI machines rather
than issues with the test logic itself.

To address this, the test is now split into four shards. Each shard is
executed separately and receives the full 15-minute timeout budget. This
should reduce the likelihood of timeout test failures.

Fixes #148384

Release note: none
Epic: none

rafiss commented:
blathers backport 25.3

backporting to resolve #152227
